README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# *c2z4uni*: a Reference Manager in *R* for universities

<p align="center">
<a 
href="https://oeysan.github.io/c2z4uni/reference/figures/badge.webp" 
id="logo" title="Logo">
<img src="man/figures/badge.webp" width="250px" alt="Logo" /> </a>
</p>
<p align="center">
<a href="https://oeysan.github.io/c2z4uni/news/index.html" id="news" 
title="News"><img src="https://img.shields.io/badge/News-2023.10.25 @ 19:01:56-purple.svg" alt="News"/></a><br/><a 
href="https://cran.r-project.org/package=c2z4uni" id="cran" 
title="CRAN Version"><img src="https://www.r-pkg.org/badges/version/c2z4uni" 
alt="CRAN Version"/></a>
<a href="https://github.com/oeysan/c2z4uni" id="github" 
title="GitHub Version"><img src="https://img.shields.io/badge/GitHub-0.1.0.9009-orange.svg" alt="GitHub Version" /></a>
<br/><a href="https://oeysan.github.io/c2z4uni/LICENSE.html" id="license" 
title="License">
<img src="https://img.shields.io/badge/Licence-MIT-blue.svg" 
alt="License" /></a>
<a href="https://github.com/oeysan/c2z4uni/actions" 
id="rcmdcheck" title="Build Status"><img 
src="https://github.com/oeysan/c2z4uni/workflows/R-CMD-check/badge.svg" 
alt="Build Status" /></a>
</p>

# 

## What is *c2z4uni* (Cristin to Zotero)?

*c2z4uni* is an extension for the [*R*](https://www.r-project.org/)
package [*c2z*](https://oeysan.github.io/c2z/). The purpose of the
package is to create a [Zotero](https://www.zotero.org/) library for
your university, department, or research group. Use with the
[Hugo](https://gohugo.io/) theme
[hugo-library](https://github.com/oeysan/hugo-library) to create an
online reference library, and/or a *R* package (e.g.,
[emayili](https://cran.r-project.org/package=emayili)) to create email
newsletters.

See [c2z4inn](https://oeysan.github.io/c2z4inn/) for a living example of
the project.

## Limitations

Please report any bugs/issues/requests
[here](https://github.com/oeysan/c2z4uni/issues/), and feel free to make
a [pull request](https://github.com/oeysan/c2z4uni/pulls).

## Code of conduct

Don’t be evil. Please read the [Code of
Conduct](https://oeysan.github.io/c2z4uni/CONDUCT.html)

## License

This project is licensed under the MIT License - see
[LICENSE](https://oeysan.github.io/c2z4uni/LICENSE.html) for details

## Acknowledgments

Henrik Karlstrøm for his work on
[rcristin](https://github.com/henrikkarlstrom/rcristin)


NEWS.md
# c2z4uni (development version)

#### Features

* `CristinMonthly` can now conduct SDG predictions derived from 
Vanderfeesten et al. (2022). Either using a self-hosted python script and models
or a user-defined host with proper using proper BERT-models

* `CristinMonthly` can now check for open source publications using `Unpaywall`
(set `get.unpaywall` to TRUE) and/or use `get.ezproxy` to append a EZproxy as
defined by `ezproxy.host`

#### Minor 

* SDG's, bibliographies, monthlies and extras should all now use key and version
to update `rds` files in `local.storage`.

* Separated `CristinMonthly` into a main function an two internal helper 
functions: `CreateMonthlies` and `CreateExtras`. `CreateMonthlies` will create
the main monthly bibliographies and store (if `local.storage` is defined) at 
`monthlies_LANG.rds`, similary `CreateExtras` will store extra metadata (e.g.,
SDG predictions, Unpaywall links, EZproxy links etc.) in `monthlies.extras.rds`

* Some minor adjustment to the internal function `InnCristin`, that now only 
searches for first and last name
* Added a missing profile image icon to user cards

#### Bug fixes

* Path to SDG images
* Workflow for windows

# c2z4uni 0.1.0.9000

* Initial launch

#### Major

* Greatly improved `CristinMonthly` (that is, making it less cumbersome) by 
speeding up the various steps. The process could be further sped up by using 
the argument `local.storage` thus pointing to a storage path of the various
outputs in rds format (possibly using sqlite in the future). The logical 
argument `full.update` (default FALSE) enables, wait for it, a full update of
collections, items, and paths.

* Moved `CristinMonthly` and `CristinUnits` from `c2z` to this package to keep
maintenance simpler.


R/cristin_json.R
#' @title Convert Cristin monthly data to JSON format
#' @description This function converts the provided Cristin monthly data into
#' JSON format, which can be used for various purposes.
#' @param cristin.monthly A list containing the monthly data from Cristin,
#' including bibliographic information.
#' @param local.storage Path to local storage directory. Default is NULL.
#' @param user.cards Logical, indicating whether to update user information
#' from INN. Default is TRUE.
#' @param full.update Logical, indicating whether to perform a full update of
#' user information from INN. Default is FALSE.
#' @param lang Language for localization ("nb", "nn", "no", or "en"). Defaults
#' to "nn".
#' @param silent Logical, indicating whether to suppress log messages. Default
#' is FALSE.
#' @param log List to store log messages. Default is an empty list.
#' @return A JSON-formatted string representing the converted Cristin monthly
#' data.
#' @details Used with `CristinMonthly` to create month-to-month bibliography in
#' JSON format
#' @examples
#' \donttest{
#'   # Create monthlies for unit 209.5.10.0
#'   example <- CristinMonthly(
#'     c2z::Zotero(
#'       id = "4827927",
#'       api = "RqlAmlH5l1KPghfCseAq1sQ1",
#'       user = FALSE
#'     ),
#'     unit.key = "209.5.10.0",
#'     start.date = "2023-07",
#'     post = TRUE,
#'     silent = TRUE
#'   )
#'
#'   # Create json data
#'   example.json <- CristinJson(example, user.cards = FALSE)
#' }
#' @rdname CristinJson
#' @export
CristinJson <- \(cristin.monthly,
                 local.storage = NULL,
                 user.cards = TRUE,
                 full.update = FALSE,
                 lang = "nn",
                 silent = FALSE,
                 log = list()) {

  # Visible bindings
  abstract <- bib.item <- key <- level <- month <- name <- parentCollection <-
    sdg <- title <- type <- year <- inn.cards <-  cristin.id <-
    cristin.ids <- ezproxy <- unpaywall <- NULL

  # Languages
  # Set language to en if not no
  lang <- if (lang %in% c("no", "nn", "nb")) "no" else "en"

  # Definitions
  monthlies <- cristin.monthly$monthlies
  unit.paths <- cristin.monthly$unit.paths

  # Define menu and link items
  menu <- tibble::tibble(
    startYear = min(monthlies$year),
    endYear = max(monthlies$year),
    nCollections = max(unit.paths$level),
    type = list(unique(monthlies$type)),
    cardURL = Dict("inn.url", lang),
    cristinURL = "https://app.cristin.no/results/show.jsf?id=",
    zoteroURL = paste0(dirname(monthlies$zotero.url[[1]]),"/")
  )

  # Define main html elements
  html <- tibble::tibble(
    bibBody = gsub("%s", "", monthlies$bib.body[[1]])
  )

  # Define collections
  collections <- unit.paths |>
    dplyr::transmute(
      key,
      level,
      name,
      parent = parentCollection
    )

  # Update user info @ inn if user.cards = TRUE
  if (user.cards) {
    inn.cards <- GetCards(local.storage, full.update, lang, silent)
  } # End user.cards

  # Function to find contributors in Inn Cards
  CreateContributors <- \(x, inn.cards = NULL) {
    if (!any(nrow(inn.cards))) {
      return (NULL)
    }
    inn.card <- inn.cards |>
      dplyr::filter(cristin.id %in% x) |>
      dplyr::arrange(match(cristin.id, x)) |>
      dplyr::pull(cristin.id)
    if (!any(length(inn.card))) inn.card <- NULL
    return (inn.card)
  }

  # Create items
  items <- monthlies |>
    dplyr::transmute(
      key,
      cristinId = cristin.id,
      type,
      html = bib.item,
      contributors = purrr::map(
        cristin.ids, ~
          CreateContributors(.x, inn.cards),
        .progress = !silent
      ),
      abstract,
      title,
      year,
      month,
      collections,
      sdg,
      unpaywall = GoFish(unpaywall),
      ezproxy = GoFish(ezproxy)
    )

  # Create json data
  json <- jsonlite::toJSON(list(
    menu = jsonlite::unbox(menu),
    html = jsonlite::unbox(html),
    collections = collections,
    items = items
  ))

  # Remove empty elements
  json <- gsub(",\"\\w+?\":(\\{\\}|\\[\\])", "", json)

  return (json)

}


R/cristin_mail.R
#' @title Create HTML Email Content from Cristin Data
#' @description This function generates HTML content for an email using data
#' from Cristin.
#' @param unit.key The key of the unit for which the email content is generated.
#' @param cristin.monthly Data containing monthly information from Cristin.
#' @param archive.host Host of the archive.
#' @param subject The subject of the email. If not provided, it will be
#' generated based on the data.
#' @param header The header of the email. If not provided, it will be generated
#' based on the data.
#' @param footer The footer of the email. If not provided, a default footer
#' will be used.
#' @param replace.style Logical, indicating whether to replace styles in the
#' generated HTML.
#' @param width Width of the email content in pixels.
#' @param lang Language for localization ("nb", "nn", "no", or "en"). Defaults
#' to "nn".
#' @param get.ezproxy Use ezproxy, Default: FALSE
#' @param ezproxy.host ezproxy host, Default: 'inn.no'
#' @return A list containing the subject and HTML body of the email.
#' @details Used with `CristinMonthly` to create month-to-month bibliography in
#' HTML format for email of selected units
#' @param silent c2z is noisy, tell it to be quiet, Default: FALSE
#' @examples
#' \donttest{
#'   # Create monthlies for unit 209.5.10.0
#'   example <- CristinMonthly(
#'     c2z::Zotero(
#'       id = "4827927",
#'       api = "RqlAmlH5l1KPghfCseAq1sQ1",
#'       user = FALSE
#'     ),
#'     unit.key = "209.5.10.0",
#'     start.date = "2023-07",
#'     post = TRUE,
#'     silent = FALSE
#'   )
#'
#'   # Create HTML email
#'   if (any(nrow(example$monthlies))) {
#'     example.mail <- CristinMail("209.5.10.0", example, "example.host")
#'
#'     # Show subject
#'     cat(example.mail$subject)
#'   }
#' }
#' @rdname CristinMail
#' @export
CristinMail <- \(unit.key,
                 cristin.monthly,
                 archive.host,
                 subject = NULL,
                 header = NULL,
                 footer = NULL,
                 replace.style = TRUE,
                 width = 700,
                 lang = "nn",
                 get.ezproxy = FALSE,
                 ezproxy.host = "inn.no",
                 silent = FALSE) {

  # Visible bindings
  body <- month.key <- year.month <- year <- id <- collections <-
    collection.names <- bib.item <- bib.body <- key <- NULL

  # Languages
  # Set language to en if not no
  lang <- if (lang %in% c("no", "nn", "nb")) "no" else "en"

  # Definitions
  monthlies <- cristin.monthly$monthlies
  unit.paths <- cristin.monthly$unit.paths


  # Function to add archive URL to publication
  AddAchive <- \(bib.item, key, lang) {

    bib.archive <- sprintf(
      " | <a href=\"%1$s/%2$s/pub/%3$s\">%4$s</a>",
      archive.host,
      lang,
      key,
      Dict("archive", lang)
    )
    bib.item <- gsub(
      "(</div>\\s*$)",
      paste0(bib.archive, "\\1"),
      bib.item,
      perl = TRUE
    )

    return (bib.item)
  }

  # Keep only latest entry
  bib <- monthlies |>
    dplyr::filter(
      grepl(
        dplyr::filter(unit.paths, id == unit.key)$key,
        collections)
    ) |>
    dplyr::group_split(year.month) |>
    rev() |>
    purrr::pluck(1) |>
    tidyr::unnest(collection.names) |>
    dplyr::arrange(
      dplyr::across(
        dplyr::starts_with("collection.name"),
        ~ data.frame(!is.na(.x), .x)
      )
    ) |>
    dplyr::mutate(
      bib.item = purrr::pmap_chr(
        list(bib.item, key, lang), ~ AddAchive(...)
      ),
      bib = purrr::pmap_chr(
        list(bib.body, bib.item), ~ sprintf(.x, .y)
      ),
      dplyr::across(
        c(year, dplyr::starts_with("collection.name")), ~
          replace(.x, duplicated(.x), NA)
      ),
      # Create headers for each row
      headers = purrr::pmap_chr(
        dplyr::across(dplyr::starts_with("collection.name")), ~ {
          x <- c(...)
          lapply(seq_along(x), \(i) {
            if (!is.na(x[[i]])) {
              if (i == 1) {
                sprintf("<h1 style=\"text-align:center;\">%s</h1>", x[[i]])
              } else {
                sprintf("<h%1$s>%2$s</h%1$s>", i, x[[i]])
              }
            }
          }) |>
            ToString("\n") |>
            GoFish()
        }
      )
    )

  # Return NULL if no data
  if (!any(nrow(bib))) {
    return (NULL)
  }

  # replace style from bibliography
  ## CSS is poorly supported in many email clients
  if (replace.style) {
    bib$bib <- gsub(
      "*.style=(\"|')[^(\"|')]*(\"|')",
      "",
      bib$bib
    )
    # Add space between items
    bib$bib <- gsub(
      "class=\"csl-entry\"",
      "class=\"csl-entry\" style=\"margin: 0.5em 0;\"",
      bib$bib
    )
  }

  # Append EZproxy if get.ezproxy = TRUE
  if (get.ezproxy) {
    bib <- bib |>
      dplyr::mutate(
        bib = purrr::map_chr(
          bib, ~ Ezproxy(.x, TRUE, ezproxy.host),
          .progress = if (!silent) "Defining EZproxy" else FALSE
        )
      )
  }

  # Define subject
  if (is.null(subject)) {
    subject <- sprintf(
      Dict("email.subject", lang),
      tolower(Month(bib$month[[1]], lang = lang)),
      bib$year[[1]]
    )
  }

  # Define header
  if (is.null(header)) {
    header <- sprintf(
      Dict("email.header", lang),
      ToString(AncestorPath(
        unit.paths,
        dplyr::filter(unit.paths, id == unit.key)$key,
        name = TRUE
      )),
      tolower(Month(bib$month[[1]], lang = lang)),
      bib$year[[1]],
      archive.host
    )
  }

  # Define footer
  if (is.null(footer)) {
    footer <- sprintf(
      "<div style=\"font-size:80%%;margin: 2em auto 0 auto; text-align: center;\">
    %s <i><a href=\"https://oeysan.github.io/c2z4uni/\">c2z4uni</a></i></div>",
      Dict("c2z", lang)
    )
  }

  # Create HTML body with references from Cristin
  body <- Trim(sprintf(
    "<table width=\"%1$s\" cellpadding=\"0\" cellspacing=\"0\">
    <tr><td>
    <div style=\"max-width:%1$spx;\">
    %2$s%3$s%4$s
    </div>
    </td></tr>
    </table>",
    width,
    header,
    ToString(Interleave(bib$headers, bib$bib), "\n"),
    footer
  ))

  return (list(subject = subject, body = body))

}


R/cristin_units.R
#' @title create a tibble with information about selected Cristin units
#' @description Create a tibble with information about (nested) units in Cristin
#'  (e.g., A University -> Faculties -> Departments -> Groups). The tibble can
#'  than be used to extract data for each unit from Cristin. Used by
#'  `CristinMonthly`
#' @param unit.key Unit to search for
#' @param subunits Add subunits, Default: TRUE
#' @param recursive Search for (nested) sub-units, Default: FALSE
#' @param ancestors Search for parent units, Default: FALSE
#' @param lang PARAM_DESCRIPTION, Default: `en`
#' @return A tibble containing information about selected units
#' @details Used with `CristinMontlhy` to create month-to-month bibliography of
#' selected units
#' @examples
#' \donttest{
#' # Find units for Inland University
#'   CristinUnits("209.0.0.0") |>
#'     dplyr::select(id, path2) |>
#'     print(width = 80)
#' }
#' @seealso
#'  \code{\link[httr]{RETRY}}
#'  \code{\link[dplyr]{transmute}}, \code{\link[dplyr]{bind_rows}}
#'  \code{\link[purrr]{pmap}}
#' @rdname CristinUnits
#' @export
CristinUnits <- \(unit.key,
                  subunits = TRUE,
                  recursive = FALSE,
                  ancestors = FALSE,
                  lang = "en") {

  # Visible bindings
  cristin_unit_id <- unit_name <- parent_unit <- name <- parent_units <-
    id <- core <- NULL

  # Languages
  # Set lang as nn if no
  if (lang %in% c("no")) lang <- "nn"
  # Set lang to en if not norwegian
  if (!lang %in% c("nb", "nn", "no")) lang <- "en"

  # Definitions
  affiliation <- Dict("affiliation", lang)

  # Find units from Cristin API
  httr.get <- Online(
    httr::RETRY(
      "GET",
      sprintf("https://api.cristin.no/v2/units/%s/", unit.key),
      query = list(lang = lang),
      quiet = TRUE
    ),
    message = "Cristin API",
    silent = TRUE
  )

  # Return NULL data if error
  if (httr.get$error) {
    return(NULL)
  }

  # Define units data
  units.data <- jsonlite::fromJSON(ParseUrl(httr.get$data))

  # Create units information
  units <- GoFish(units.data$parent_units$unit_name, NULL) |>
    c(name = units.data$unit_name) |>
    unlist() |>
    bind_rows() |>
    dplyr::mutate(
      path = dplyr::case_when(
        "subunits" %in% names(units.data) ~ affiliation,
        TRUE ~ NA_character_
      ),
      core = purrr::pmap(dplyr::across(!dplyr::last_col()), ~ unname(c(...)))
    ) |>
    dplyr::select_if(~ !all(is.na(.))) |>
    dplyr::rename_with(~ paste0("path", seq_along(.)), !dplyr::last_col()) |>
    tibble::add_column(id = units.data$cristin_unit_id, .before = 1)

  # Find and return unit with ancestors if ancestors = TRUE
  if (ancestors) {
    unit.ids <- c(
      GoFish(units.data$parent_units$cristin_unit_id, NULL),
      units.data$cristin_unit_id
    )
    if (length(unit.ids) > 1) {
      units <- CristinUnits(unit.ids[[1]], recursive = TRUE, lang = lang) |>
        dplyr::filter(id %in% unit.ids)
    }
  }

  # Return if !subunits | "subunits" %in% names(units.data)
  if (!subunits | !"subunits" %in% names(units.data)) {
    return (units)
  }

  # Else add subunits
  units <- dplyr::bind_rows(
    units,
    units.data$subunits |>
      tibble::as_tibble() |>
      dplyr::transmute(
        id = cristin_unit_id,
        !!paste0("path", ncol(units)-2) := unlist(unit_name)
      )
  ) |>
    tidyr::fill(dplyr::starts_with("path"))

  # Add nested sub units if recursive is TRUE
  if (recursive) {
    units <- dplyr::bind_rows(
      units[1, ],
      lapply(seq_len(nrow(units))[-1], \(i) {
        CristinUnits(
          units[i, ]$id,
          subunits = TRUE,
          recursive = TRUE,
          lang = lang)
      })
    )
  }

  # Arrange units
  units <- units |>
    dplyr::select(c(id, core, dplyr::starts_with("path"))) |>
    dplyr::arrange(
      dplyr::across(
        dplyr::starts_with("path"),
        ~ data.frame(!is.na(.x), !.x %in% affiliation, .x)
      )
    )

  return(units)

}


R/cristin_web.R
#' @title Create HTML and Metadata Files from Cristin Data for Web Display
#' @description This function generates HTML and metadata files for web
#' display using data from Cristin.
#' @param monthlies Data containing monthly information from Cristin.
#' @param sdg.data Data containing information about Sustainable Development
#' Goals (SDGs). Default is NULL.
#' @param sdg.path Path to SDG images. Default is NULL.
#' @param archive.url Relative URL to archive. Default is NULL.
#' @param local.storage Path to local storage directory. Default is NULL.
#' @param user.cards Logical, indicating whether to update user information
#' from INN. Default is TRUE.
#' @param full.update Logical, indicating whether to perform a full update of
#' user information from INN. Default is FALSE.
#' @param lang Language for localization ("nb", "nn", "no", or "en"). Defaults
#' to "nn".
#' @param silent Logical, indicating whether to suppress log messages. Default
#' is FALSE.
#' @param log List to store log messages. Default is an empty list.
#' @return A tibble containing metadata and HTML content for web display.
#' @details Used with `CristinMonthly` to create month-to-month bibliography in
#' markdown format
#' @examples
#' \donttest{
#'   # Create monthlies for unit 209.5.10.0
#'   example <- CristinMonthly(
#'     c2z::Zotero(
#'       id = "4827927",
#'       api = "RqlAmlH5l1KPghfCseAq1sQ1",
#'       user = FALSE
#'     ),
#'     unit.key = "209.5.10.0",
#'     start.date = "2023-07",
#'     post = TRUE,
#'     silent = TRUE
#'   )
#'
#'   # Create md files
#'   example.web <- CristinWeb(
#'     example$monthlies,
#'     user.cards = FALSE
#'   )
#'
#'   # Print the three first individual pages (if any)
#'   if (!is.null(example$monthlies)) {
#'     example$monthlies |>
#'       dplyr::select(title) |>
#'       print(n = 3, width = 80)
#'   }
#' }
#' @rdname CristinWeb
#' @export
CristinWeb <- function(monthlies,
                       sdg.data = NULL,
                       sdg.path = NULL,
                       archive.url = NULL,
                       local.storage = NULL,
                       user.cards = TRUE,
                       full.update = FALSE,
                       lang = "nn",
                       silent = FALSE,
                       log = list()) {

  # Visible bindings
  cristin.id <- inn.cards <- NULL

  # Languages
  # Set language to en if not no
  lang <- if (lang %in% c("no", "nn", "nb")) "no" else "en"

  # Update user info @ inn if user.cards = TRUE
  if (user.cards) {
    inn.cards <- GetCards(local.storage, full.update, lang, silent)
  } # End user.cards

  # Function to create md files
  CreateMd <- function(item) {
    # Visible bindings
    title <- reference <- abstract <- contributors <- sdg <- archive <-
      unpaywall <- ezproxy <- reference.button <- abstract.button <-
      contributors.button <- sdg.button <- cristin.button <- zotero.button <-
      archive.button <- unpaywall.button <- ezproxy.button <- NULL

    # Function to create a vector of collection names
    CollectionNames <- function(x) {
      x |>
        unlist() |>
        strsplit(split = "\\|\\|") |>
        unlist() |>
        as.character() |>
        Trim()
    }

    # Define title
    if (any(!is.na(GoFish(item$title)))) {
      title <- item$title
    } else {
      title <- item$key
    }

    # Reference
    if (any(!is.na(GoFish(item$bib)))) {
      reference <- htmltools::tags$article(
        id = paste0("csl-bib-container-", item$key),
        class = "csl-bib-container",
        htmltools::HTML(item$bib)
      )
    }

    # Abstract
    if (any(!is.na(GoFish(item$abstract)))) {
      abstract <- htmltools::tags$article(
        htmltools::h1(Dict("abstract", lang)),
        item$abstract,
        id = paste0("abstract-article-", item$key),
        class = "abstract-article"
      )
      abstract.button <- htmltools::a(
        Dict("abstract", lang),
        href = paste0("#", paste0("abstract-article-", item$key)),
        class = "csl-bib-button"
      )
    }

    # Contributors
    if (any(!is.na(GoFish(item$cristin.ids[[1]]))) &
        any(nrow(inn.cards))) {

      ids <- item$cristin.ids[[1]]
      inn.card <- inn.cards |>
        dplyr::filter(cristin.id %in% ids) |>
        dplyr::arrange(match(cristin.id, ids))

      if (any(nrow(inn.card))) {
        contributors <- htmltools::tags$article(
          htmltools::h1(Dict("contributors", lang)),
          htmltools::HTML(inn.card$card),
          id = paste0("contributors-article-", item$key),
          class = "contributors-article"
        )
        contributors.button <- htmltools::a(
          Dict("contributors", lang),
          href = paste0("#", paste0("contributors-article-", item$key)),
          class = "csl-bib-button"
        )
      }
    }

    # SDG
    if (any(!is.na(GoFish(item$sdg[[1]]))) & any(length(sdg.data$sum))) {

      sdg <- htmltools::tagAppendChild(
        htmltools::tags$article(
          htmltools::h1(Dict("sdg", lang)),
          id = paste0("sdg-article-", item$key),
          class = "sdg-article"
        ),
        htmltools::div(
          SdgInfo(
            sdg.data$sum,
            as.numeric(item$sdg[[1]]),
            lang = lang,
            sdg.path,
            archive.url
          ) |>
            do.call(what = htmltools::HTML),
          class = "sdg-container"
        )
      )

      sdg.button <- htmltools::a(
        Dict("sdg", lang),
        href = paste0("#", paste0("sdg-article-", item$key)),
        class = "csl-bib-button"
      )
    }

    # Archive
    tags <- c(
      CollectionNames(item$collection.names),
      item$year,
      Month(item$month, lang)
    )
    archive <- htmltools::tags$article(
      htmltools::h1(Dict("archive", lang)),
      htmltools::tags$ul(
        purrr::map(tags, ~ htmltools::tags$li(.x))
      ),
      id = paste0("taxonomy-article-", item$key),
      class = "taxonomy-article"
    )
    archive.button <- htmltools::a(
      Dict("archive", lang),
      href = paste0("#", paste0("taxonomy-article-", item$key)),
      class = "csl-bib-button"
    )

    # Unpaywall
    if (any(!is.na(GoFish(item$unpaywall)))) {
      unpaywall.button <- htmltools::a(
        "Unpaywall",
        href = item$unpaywall,
        class = "csl-bib-button"
      )
    }

    # Unpaywall
    if (any(!is.na(GoFish(item$ezproxy)))) {
      ezproxy.button <- htmltools::a(
        "EZproxy",
        href = item$ezproxy,
        class = "csl-bib-button"
      )
    }

    # Add Cristin and Zotero buttons
    cristin.button <- htmltools::a(
      "Cristin",
      href = item$cristin.url,
      alt = "Cristin URL",
      class = "csl-bib-button"
    )
    zotero.button <- htmltools::a(
      "Zotero",
      href = item$zotero.url,
      alt = "Zotero URL",
      class = "csl-bib-button"
    )

    # Combine params
    md.frontmatter <- sprintf(
      'title: "%s"\ntype: pub',
      gsub("\"", "'", title)
    )

    md.html <- htmltools::tagList(
      htmltools::h1(Dict("publications", lang)),
      htmltools::tagAppendChildren(
        # Reference
        reference,
        # Buttons
        htmltools::div(
          archive.button,
          cristin.button,
          zotero.button,
          abstract.button,
          contributors.button,
          sdg.button,
          unpaywall.button,
          ezproxy.button,
          class = "csl-bib-buttons"
        ),
        htmltools::div(
          id = paste0("csl-bib-meta-container-", item$key)
        )
      ),
      htmltools::div(
        # Abstract
        abstract,
        # Contributors
        contributors,
        # SDG
        sdg,
        # Taxonomy
        archive,
        id = paste0("csl-bib-meta-", item$key),
        class = "csl-bib-meta"
      )
    )

    md <- tibble::tibble(
      key = item$key,
      frontmatter = md.frontmatter,
      html = as.character(md.html),
      year.month = item$year.month
    )

    return (md)
  }

  # Markdownlist
  markdowns <- list()
  # Start time for query
  query.start <- Sys.time()
  # Cycle through monthlies and create markdowns
  for (i in seq_len(nrow(monthlies))) {
    markdowns[[i]] <- CreateMd(monthlies[i, ])
  # Estimate time of arrival
  log.eta <-
    LogCat(
      Eta(query.start,
          i,
          nrow(monthlies)),
      silent = silent,
      flush = TRUE,
      log = log,
      append.log = FALSE
    )
  }

  # Create tibble
  markdowns <- dplyr::bind_rows(markdowns)

  return (markdowns)
}
